lib/replicache.dart

```diff
@@ -192,8 +230,102 @@ class Replicache implements ReadTransaction {
   }) =>
       query((tx) => tx.scan(prefix: prefix, start: start, limit: limit));
 
-  /// Synchronizes the database with the server. New local transactions that have been executed since the last
-  /// sync are sent to the server, and new remote transactions are received and replayed.
+  Future<void> _sync() async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final syncHead = await _beginSync();
+    await _maybeEndSync(syncHead);
+  }
+
+  Future<String> _beginSync() async {
+    await _opened;
+    final beginSyncResult = await _invoke(_name, 'beginSync', {
+      'batchPushURL': _batchUrl,
+      // TODO(arv): Rename this param.
+      'clientViewURL': _remote,
+      'dataLayerAuth': _clientViewAuth,
+    });
+
+    final syncInfo = beginSyncResult['syncInfo'];
+
+    void checkStatus(Map<String, dynamic> data, String serverName) {
+      final httpStatusCode = data['httpStatusCode'];
+      switch (httpStatusCode) {
+        case HttpStatus.ok:
+          break;
+        case HttpStatus.unauthorized:
+          // TODO: Reauth
+          break;
+        default:
+          print(
+              'Got unexpected response from $serverName server: $httpStatusCode: ${data['errorMessage']}');
+      }
+    }
+
+    final batchPushInfo = syncInfo['batchPushInfo'];
+    if (batchPushInfo != null) {
+      checkStatus(batchPushInfo, 'batch');
+      final mutationInfos = batchPushInfo['batchPushResponse']['mutationInfos'];
+      if (mutationInfos != null) {
+        for (final mutationInfo in mutationInfos) {
+          print(
+              'MutationInfo: ID: ${mutationInfo['id']}, Error: ${mutationInfo['error']}');
+        }
+      }
+    }
+
+    checkStatus(syncInfo['clientViewInfo'], 'client view');
+
+    return beginSyncResult['syncHead'];
+  }
+
+  Future<void> _maybeEndSync(String syncHead) async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final res = await _invoke(_name, 'maybeEndSync', {'syncHead': syncHead});
+    final replayMutations = res['replayMutations'];
+    if (replayMutations == null || replayMutations.isEmpty) {
```

Comment:
if replay mutations is null/empty that is not sufficient to say that sync completed successfully. we also need to check that maybeendsync did not return an error because replaymutations will be empty in that case too. i dont know dart and dont understand how the error comes back from https://github.com/rocicorp/replicache-client/blob/8a2d797818bdb97b6ebbd0e2538d167c5408562b/repm/api.go#L181 so maybe this is handled in some way i dont see.

Reply:
The errors from Go get translated into exceptions by lower-level machinery. `sync` catches any exceptions at all and re-schedules itself.